gateway/src/apicast/http_proxy.lua

```diff
@@ -96,10 +97,24 @@ local function forward_https_request(proxy_uri, uri)
         -- We cannot use resty.http's .get_client_body_reader().
         -- In POST requests with HTTPS, the result of that call is nil, and it
         -- results in a time-out.
+        --
+        --
+        -- If ngx.req.get_body_data is nil, can be that the body is too big to
+        -- read and need to be cached in a local file. This request will return
+        -- nil, so after this we need to read the temp file.
+        -- https://github.com/openresty/lua-nginx-module#ngxreqget_body_data
         body = ngx.req.get_body_data(),
         proxy_uri = proxy_uri
     }
 
+    if not request.body then
+        local temp_file_path = ngx.req.get_body_file()
+        ngx.log(ngx.INFO, "HTTPS Proxy: Request body is too big, read the content from path='", temp_file_path, "'")
+        if temp_file_path then
+          request.body = util.read_file(temp_file_path)
```

Comment:
This is not safe for big files, because it loads everything in memory. It needs to be an iterator that reads the body chunk by chunk.

Reply:
Hi,

I'm not sure about this, but Lua already make this on read_all option:

Call:
https://github.com/lua/lua/blob/61a4e64a6667bedaa882571c48a173ef5a4ba73b/liolib.c#L583

Read_all
https://github.com/lua/lua/blob/61a4e64a6667bedaa882571c48a173ef5a4ba73b/liolib.c#L526-L538

So, I'm not sure what do you want me to do here.

Reply:
But `apicast.util.read_file` is https://github.com/3scale/APIcast/blob/14df1a809fd422809fc1c3de9e44f82e6a2a5ada/gateway/src/apicast/util.lua#L25-L37

Right?
It reads the whole file into memory.

It should not read the whole file in memory, but rather return an iterator that reads each chunk and returns it. 

Like https://github.com/ledgetech/lua-resty-http#get_client_body_reader. Actually, that looks exactly like that.

Reply:
Hey,

I was looking at this, but at some point, the information need to be saved in
memmory, maybe I'm missing something, but let explain the body reader that you sent to me:


There are two functions involved:

- **_body_reader**: This is using coroutines, and this is correct, all here
  looks good, it returns a iterator and all is ok

https://github.com/ledgetech/lua-resty-http/blob/a6bd2e0eb1390e330e4fb10a48cced5a1f21fb66/lib/resty/http.lua#L444-L505


- **_read_body**: In this case, when reader() is called, all the information is
  saved in a table, and after that is all in a single variable. This, in my
  opinion, is a bit worse, because you're allocate a table instead of append of
  the string (But I need to do a performance check here ;-), not a strong word)

https://github.com/ledgetech/lua-resty-http/blob/a6bd2e0eb1390e330e4fb10a48cced5a1f21fb66/lib/resty/http.lua#L512-L536


So, I was thinking to have a metatable with some kind of magic method, but
AFAIK, always the content should be stored in memory. I was thinking something
like Filereader in Golang, but I didn't found a way to do that.

Reply:
Nginx buffers large bodies (more than a few kb) into files, so they don't need to be stored in memory. 
Reading the *whole* file into memory can lead to DoS as it can have megabytes or gigabytes.
So a reading iterator will read only a chunk of the file at the time and iterating over it.

I have no idea if the iterator I linked would actually work when the body is buffered into a file. It was a mere suggestion of what should happen when reading a file from a filesystem - read it in chunks as an iterator. Maybe penlight has something?